CMakeLists.txt

```diff
@@ -808,7 +808,7 @@ if(WITH_OPTIM)
         endif()
         if(WITH_SSSE3)
             check_ssse3_intrinsics()
-            if(HAVE_SSSE3_INTRIN)
+            if(HAVE_SSSE3_INTRIN AND WITH_SSE2)
```

Comment:
This fifth commit confuses me. It adds explicit dependencies that are now made redundant by one of the earlier commits [Fix x86 WITH_* options dependencies](https://github.com/zlib-ng/zlib-ng/pull/1643/commits/ccdaab5e09404732047f945fd1f03e242b615c26).

In this line, `WITH_SSE2` will always be true because `WITH_SSSE3` cannot be true without it.
The same kind of thing repeats for many of the changes in this commit.

I am confused, what is the rationale behind these changes?

Reply:
``cmake_dependent_option`` specifies dependencies for build options. We can request support for an instruction set that is not supported by the compiler.

This change defines dependencies for compiler-supported intrinsics. If the compiler does not support ``SSE2`` intrinsics, we disable ``SSSE3``.

See https://github.com/zlib-ng/zlib-ng/pull/1643/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR806

If the compiler does not support ``SSE2`` we set ``WITH_SSE2`` to ``OFF``.

Reply:
Then doesn't this `AND WITH_SSE2` check really belong two lines above, at line 809?
There is no reason to run `check_ssse3_intrinsics()` if its results always gets ignored because of the check on the next line.. I don't see any other reason to have the check where it is, unless you hope to add non-intrinsic SSSE3 support, but that would require ASM, and we try to avoid that.

Reply:
Only for simplification.

My code (we need set ``WITH_SSSE3`` to ``OFF`` if ``WITH_SSE2`` is ``OFF`` or compiler not support ``SSSSE3``):

```cmake
        if(WITH_SSSE3)
            check_ssse3_intrinsics()
            if(HAVE_SSSE3_INTRIN AND WITH_SSE2)
                add_definitions(-DX86_SSSE3)
                set(SSSE3_SRCS ${ARCHDIR}/adler32_ssse3.c ${ARCHDIR}/chunkset_ssse3.c)
                add_feature_info(SSSE3_ADLER32 1 "Support SSSE3-accelerated adler32, using \"${SSSE3FLAG}\"")
                list(APPEND ZLIB_ARCH_SRCS ${SSSE3_SRCS})
                set_property(SOURCE ${SSSE3_SRCS} PROPERTY COMPILE_FLAGS "${SSSE3FLAG} ${NOLTOFLAG}")
            else()
                set(WITH_SSSE3 OFF)
            endif()
        endif()
```

Your suggestion:

```cmake
        if(WITH_SSSE3 AND WITH_SSE2)
            check_ssse3_intrinsics()
            if(HAVE_SSSE3_INTRIN)
                add_definitions(-DX86_SSSE3)
                set(SSSE3_SRCS ${ARCHDIR}/adler32_ssse3.c ${ARCHDIR}/chunkset_ssse3.c)
                add_feature_info(SSSE3_ADLER32 1 "Support SSSE3-accelerated adler32, using \"${SSSE3FLAG}\"")
                list(APPEND ZLIB_ARCH_SRCS ${SSSE3_SRCS})
                set_property(SOURCE ${SSSE3_SRCS} PROPERTY COMPILE_FLAGS "${SSSE3FLAG} ${NOLTOFLAG}")
            else()
                set(WITH_SSSE3 OFF)
            endif()
        else()
            set(WITH_SSSE3 OFF)
        endif()
```

Yes, ``check_ssse3_intrinsics`` will be called unnecessarily, but I think it simplifies the code.